ev3dev/core.py

```diff
@@ -2615,27 +2610,130 @@ def __init__(self, address=None, name_pattern=SYSTEM_DEVICE_NAME_CONVENTION, nam
     #: Calibration ???
     MODE_IR_CAL = 'IR-CAL'
 
-
     MODES = (
-      'IR-PROX',
-      'IR-SEEK',
-      'IR-REMOTE',
-      'IR-REM-A',
-      'IR-CAL',
+      MODE_IR_PROX,
+      MODE_IR_SEEK,
+      MODE_IR_REMOTE,
+      MODE_IR_REM_A,
+      MODE_IR_CAL
     )
 
+    # The following are all of the various combinations of button presses for
+    # the remote control.  The key/index is the number that will be written in
+    # the attribute file to indicate what combination of buttons are currently
+    # pressed.
+    _BUTTON_VALUES = {
+            0: [],
+            1: ['red_up'],
+            2: ['red_down'],
+            3: ['blue_up'],
+            4: ['blue_down'],
+            5: ['red_up', 'blue_up'],
+            6: ['red_up', 'blue_down'],
+            7: ['red_down', 'blue_up'],
+            8: ['red_down', 'blue_down'],
+            9: ['beacon'],
+            10: ['red_up', 'red_down'],
+            11: ['blue_up', 'blue_down']
+            }
+
+    #: Handles ``Red Up`` events.
+    on_red_up = None
+
+    #: Handles ``Red Down`` events.
+    on_red_down = None
```

Comment:
These are user-defined event handling functions that should fire up during call to `process()` method (and by the way, I don't see the method here). As in
```py
def beep_when_pressed(state):
    if state:
      Sound.beep()

rc.on_red_up = beep_when_pressed

while True:
  rc.process()
```

I don't see how these would work as is, since we are merging `RemoteControl` with the `InfraredSensor`, and there is no way to set the channel here (we used to set channel in the constructor of `RemoteControl`).

We should probably change these into dictionaries: `on_press[1] = beep_when_pressed` (not lists because channel numbers start at 1).

Reply:
Re `process()` method: in the `RemoteControl` class it was inherited from `ButtonBase`. If we do want to merge `RemoteControl` with `InfraredSensor`, we should probably inherit from `ButtonBase` as well.

Reply:
gotcha, will test/work on this part this evening

Reply:
This is working now

Reply:
How exactly does it work now? Where do we specify channel to use with the handles?

Reply:
The `channel` has to be passed to the `process()` call.  Here is `RemoteControlledTank` from `helper.py`
```
class RemoteControlledTank(MoveTank):

    def __init__(self, left_motor_port, right_motor_port, polarity='inversed', speed=400, channel=1):
        MoveTank.__init__(self, left_motor_port, right_motor_port)
        self.set_polarity(polarity)

        left_motor = self.motors[left_motor_port]
        right_motor = self.motors[right_motor_port]
        self.speed_sp = speed
        self.remote = InfraredSensor()
        self.remote.on_red_up = self.make_move(left_motor, self.speed_sp)
        self.remote.on_red_down = self.make_move(left_motor, self.speed_sp* -1)
        self.remote.on_blue_up = self.make_move(right_motor, self.speed_sp)
        self.remote.on_blue_down = self.make_move(right_motor, self.speed_sp * -1)
        self.channel = channel

    def make_move(self, motor, dc_sp):
        def move(state):
            if state:
                motor.run_forever(speed_sp=dc_sp)
            else:
                motor.stop()
        return move

    def main(self):

        try:
            while True:
                self.remote.process(self.channel)
                time.sleep(0.01)

        # Exit cleanly so that all motors are stopped
        except (KeyboardInterrupt, Exception) as e:
            log.exception(e)
            self.stop()
```

Reply:
Speaking of `helper.py`, we are not too far from being able to get rid of it
- All of the motor functionality that I put in there now exists in core.py (it didn't at the time that I wrote it).  I need to update MINDCUB3R to not use the helper.py and then all of the motor code in there can go.
- The wheel classes could be moved into a `wheel.py`, BALANC3R and GYROBOY demos will need to be updated.
- We need to find a home for `RemoteControlledTank`
- `ColorSensorMixin` only added the `rgb()` call and there weren't any objections to adding that method to core.py

Reply:
> The channel has to be passed to the process() call. 

So we loose the ability to have different handlers for different channels? I understand that this is the restriction of how `ButtonBase.process()` is implemented, but it seems an important feature (why have several channels, when they behave in the same way?). :thinking:

Reply:
> So we loose the ability to have different handlers for different channels?

We did not have that capability before either. Prior to this PR you would need to create a RemoteControl object for each channel that you wished to define handlers for. With this PR you need to create an InfraredSensor object for each channel that you wish to define handlers for.

If we think that supporting per-channel handlers in a single object is something folks will use let's add that via another PR.  It shouldn't be too bad, we just have the handler mappings be a dictionary where the channel is the key.

Reply:
> Prior to this PR you would need to create a RemoteControl object for each channel that you wished to define handlers for. 

Yes, that is correct. But the new implementation of infrared sensor allows to (or claims to) be channel-independent: channel is passed to methods like `heading()` or `process()` instead of being assigned at construction. But for RC handlers to work with different channels users would have to create an instance of `InfraredSensor` per channel, which does not seem logical. I think we could drop `ButtonBase` from parents, and reimplement the functionality so that handlers would take not just state, but also channel, or something to the same effect?

Reply:
> If we think that supporting per-channel handlers in a single object is something folks will use 

I used this ability for the big shooting robot (EV3STORM?). One channel for leg movement, the other for shooting. It does come handy from time to time.

Reply:
Good point. Are you ok if we merge this PR and file a new issue to add per-channel capabilities to the handlers?  We are at 10 commits and 60+ comments on this PR and short of the per-channel handlers I think it is in pretty good shape.

Reply:
Let's do that, it seems in good shape otherwise.

Reply:
Filed #380 for this part